bed/enrichment.go

```diff
@@ -0,0 +1,232 @@
+
+	answer[0] = 1.0
+	answer[1] = mu //mu represents the expected value
+
+	//calculate pValue approximation
+	pValue := numbers.NormalDist(float64(overlapCount), mu, sigma)
+	for s := overlapCount + 1; s <= len(elementOverlapProbs); s++ {
+		pValue += numbers.NormalDist(float64(s), mu, sigma)
+	}
+	answer[2] = pValue
+
+	return answer
+}
+
+//EnrichmentPValueExact performs an exact calculation fo enrichment bgaased on a set of overlap probabilities and the observed overlap count.
+//The exact method is non-polynomial, and is thus not recommended for large datasets.
+//Returns a slice of three values. The first is the debug check, the second is the expected number of overlaps, and the third is the pValue of the observed number of overlaps.
+func EnrichmentPValueExact(elementOverlapProbs []float64, overlapCount int) []float64 {
+	var numTrials = len(elementOverlapProbs)
+	var answer []float64 = make([]float64, 3)
+	var prevCol []float64 = make([]float64, numTrials+1)
+	var currCol []float64 = make([]float64, numTrials+1)
+
+	//for one trial (a bed input with a single element)
+	prevCol[0] = math.Log(1 - elementOverlapProbs[0])
+	currCol[0] = prevCol[0]
+	prevCol[1] = math.Log(elementOverlapProbs[0])
+	currCol[1] = math.Log(elementOverlapProbs[0])
+
+	var s int
+	var expected, pValue float64
+
+	for t := 1; t < numTrials; t++ {
+		prevCol, currCol = currCol, prevCol
+		currCol[0] = prevCol[0] + math.Log(1-elementOverlapProbs[t])
+		for s = 1; s <= t; s++ {
+			currCol[s] = numbers.AddLog(prevCol[s]+math.Log(1-elementOverlapProbs[t]), prevCol[s-1]+math.Log(elementOverlapProbs[t]))
+		}
+		currCol[t+1] = prevCol[t] + math.Log(elementOverlapProbs[t])
+	}
+
+	//Now we check whether our probabilities add to 1.
+	check := currCol[0]
+	for s = 1; s <= numTrials; s++ {
+		check = numbers.AddLog(check, currCol[s])
+		if s == 1 {
+			expected = currCol[s]
+		} else {
+			expected = numbers.AddLog(expected, currCol[s]+math.Log(float64(s)))
+		}
+	}
+
+	pValue = currCol[overlapCount]
+	for s = overlapCount + 1; s <= numTrials; s++ {
+		pValue = numbers.AddLog(pValue, currCol[s])
+	}
+
+	answer[0] = math.Exp(check)
+	answer[2] = math.Exp(pValue)
+	answer[1] = math.Exp(expected)
+	return answer
+}
+
+//EnrichmentPValueUpperBound, together with EnrichmentPValueLowerBound, provide a range of possible values for the pValue of overlap.
+//Returns a slice of three values. The first is the debug check, the second is the expected number of overlaps, and the third is the pValue of the observed number of overlaps.
+func EnrichmentPValueUpperBound(elements1 []Bed, elements2 []Bed, noGapRegions []Bed, overlapCount int, verbose int) []float64 {
+	var numTrials int = len(elements2)
+	var answer []float64 = make([]float64, 3)
+	var tempElements1 []Bed = make([]Bed, len(elements1))
+	copy(tempElements1, elements1)
+	var tempNoGap []Bed = make([]Bed, len(noGapRegions))
+	copy(tempNoGap, noGapRegions)
+	minElements2 := findLargestBedLength(elements2)
```

Comment:
Is this line the only difference between upper and lower bound?  Can we have a single function that takes in a single element length and number of elements?